spatialdata/_core/_spatialdata.py

```diff
@@ -591,62 +695,177 @@ def add_shapes(
-            root = zarr.group(store=store)
-            store.close()
-
-        self.path = file_path
+        if isinstance(file_path, str):
+            file_path = Path(file_path)
+        assert isinstance(file_path, Path)
+
+        if self.is_backed() and self.path != file_path:
+            logger.info(f"The Zarr file used for backing will now change from {self.path} to {file_path}")
+
+        # old code to support overwriting the backing file
+        # target_path = None
+        # tmp_zarr_file = None
+        if os.path.exists(file_path):
+            if parse_url(file_path, mode="r") is None:
+                raise ValueError(
+                    "The target file path specified already exists, and it has been detected not not be "
+                    "a Zarr store. Overwriting non-Zarr stores is not supported to prevent accidental "
+                    "data loss."
+                )
+            if not overwrite and self.path != str(file_path):
+                raise ValueError("The Zarr store already exists. Use overwrite=True to overwrite the store.")
+            elif str(file_path) == self.path:
+                raise ValueError(
+                    "The file path specified is the same as the one used for backing. "
+                    "Overwriting the backing file is not supported to prevent accidental data loss."
+                    "We are discussing how to support this use case in the future, if you would like us to "
+                    "support it please leave a comment on https://github.com/scverse/spatialdata/pull/138"
+                )
+                # old code to support overwriting the backing file
+                # else:
+                #     target_path = tempfile.TemporaryDirectory()
+                #     tmp_zarr_file = Path(target_path.name) / "data.zarr"
+
+        # old code to support overwriting the backing file
+        # if target_path is None:
+        #     store = parse_url(file_path, mode="w").store
+        # else:
+        #     store = parse_url(tmp_zarr_file, mode="w").store
+        # store = parse_url(file_path, mode="w").store
+        # root = zarr.group(store=store)
+        store = parse_url(file_path, mode="w").store
+
+        root = zarr.group(store=store, overwrite=overwrite)
+        store.close()
+
+        # old code to support overwriting the backing file
+        # if target_path is None:
+        #     self.path = str(file_path)
+        # else:
+        #     self.path = str(tmp_zarr_file)
+        self.path = str(file_path)
         try:
             if len(self.images):
                 root.create_group(name="images")
-                # add_image will delete and replace the same key in self.images, so we need to make a copy of the
+                # add_image_in_memory will delete and replace the same key in self.images, so we need to make a copy of the
                 # keys. Same for the other elements
                 keys = list(self.images.keys())
```

Comment:
it looks like we are not checking that the keys match though right>? shall we be checking for that? otherwise why make a copy?

Reply:
I didn't get exactly what you mean but I fixed this. The copy was to avoid this error: `RuntimeError: dictionary keys changed during iteration`, since `_add_image_in_memory()` (same for the other elements), was doing a `del imags[key]` when `overwrite=True`. I removed this `del` since when we do `image[key] = new_image`, we remove the old content and we don't change the keys.